scene/main/viewport.cpp

```diff
@@ -1184,56 +1182,68 @@ void Viewport::_gui_show_tooltip() {
-	Control *tooltip_owner = nullptr;
+	Control *tooltip_serve_target = nullptr;
 	String tooltip_text = _gui_get_tooltip(
 			gui.tooltip_control,
 			gui.tooltip_control->get_global_transform().xform_inv(gui.last_mouse_pos),
-			&tooltip_owner);
+			&tooltip_serve_target);
 	tooltip_text = tooltip_text.strip_edges();
 	if (tooltip_text.is_empty()) {
 		return; // Nothing to show.
 	}
 
-	// Remove previous popup if we change something.
-	if (gui.tooltip_popup) {
-		memdelete(gui.tooltip_popup);
-		gui.tooltip_popup = nullptr;
-		gui.tooltip_label = nullptr;
+	// Remove previous custom tooltip if we change something.
+	if (gui.custom_tooltip) {
+		memdelete(gui.custom_tooltip);
+		gui.custom_tooltip = nullptr;
 	}
 
-	if (!tooltip_owner) {
+	if (!tooltip_serve_target) {
 		return;
 	}
 
-	// Popup window which houses the tooltip content.
-	PopupPanel *panel = memnew(PopupPanel);
-	panel->set_theme_type_variation(SNAME("TooltipPanel"));
-
-	// Controls can implement `make_custom_tooltip` to provide their own tooltip.
-	// This should be a Control node which will be added as child to a TooltipPanel.
-	Control *base_tooltip = tooltip_owner->make_custom_tooltip(tooltip_text);
+	if (!gui.tooltip_popup) {
+		// Popup window which houses the tooltip content.
+		gui.tooltip_popup = memnew(PopupPanel);
+		gui.tooltip_popup->set_theme_type_variation(SNAME("TooltipPanel"));
+		gui.tooltip_popup->set_transient(true);
+		gui.tooltip_popup->set_flag(Window::FLAG_NO_FOCUS, true);
+		gui.tooltip_popup->set_flag(Window::FLAG_POPUP, false);
+		gui.tooltip_popup->set_wrap_controls(true);
 
-	// If no custom tooltip is given, use a default implementation.
-	if (!base_tooltip) {
+		// Provides a default implementation in case there is no custom tooltip.
 		gui.tooltip_label = memnew(Label);
 		gui.tooltip_label->set_theme_type_variation(SNAME("TooltipLabel"));
-		gui.tooltip_label->set_auto_translate(gui.tooltip_control->is_auto_translating());
-		gui.tooltip_label->set_text(tooltip_text);
-		base_tooltip = gui.tooltip_label;
-		panel->connect("mouse_entered", callable_mp(this, &Viewport::_gui_cancel_tooltip));
+		gui.tooltip_label->set_anchors_and_offsets_preset(Control::PRESET_FULL_RECT);
+		gui.tooltip_label->hide();
+		gui.tooltip_popup->add_child(gui.tooltip_label);
+		add_child(gui.tooltip_popup);
 	}
 
-	base_tooltip->set_anchors_and_offsets_preset(Control::PRESET_FULL_RECT);
-
-	panel->set_transient(true);
-	panel->set_flag(Window::FLAG_NO_FOCUS, true);
-	panel->set_flag(Window::FLAG_POPUP, false);
-	panel->set_wrap_controls(true);
-	panel->add_child(base_tooltip);
-
-	gui.tooltip_popup = panel;
+	// Because gui.tooltip_popup needs to be reused, it may already exist and only tooltip_serve_target has changed.
+	// In order to use the theme associated with tooltip_serve_target, the theme property needs to be propagated manually.
+	gui.tooltip_popup->propagate_theme_changed(tooltip_serve_target->get_theme_owner_node());
```

Comment:
This breaks the expected theme propagation logic, and is not reliable. I'd rather we didn't do that. If we need a system for a controlled theme owner node, then we need to implement it in the `ThemeOwner`.

That said, I'm not sure there are currently any requests to make the tooltip control to indirectly inherit the spawner-Control's theme.

Reply:
> That said, I'm not sure there are currently any requests to make the tooltip control to indirectly inherit the spawner-Control's theme.

It could be relevant for the editor, as we've had multiple issues/regressions over the years where tooltips would not properly use the editor theme. They should all be worked around by making it explicit in each case, but the bugs happened because we seemed to assume it would be automatic.

For games though, I don't know what would be the expectation.

Reply:
Previously tooltip popup was added as a child of `tooltip_serve_target` (previously `tooltip_owner`). It is now added as a child of the viewport for reuse.

So if the theme properties are not re-propagated, there may be a situation similar to the following figure in the editor.

![1](https://user-images.githubusercontent.com/30386067/191243965-a42fb72c-9cc6-40c8-a33b-22194d3141e3.png)

Reply:
> It could be relevant for the editor, as we've had multiple issues/regressions over the years where tooltips would not properly use the editor theme. They should all be worked around by making it explicit in each case, but the bugs happened because we seemed to assume it would be automatic.

I actually have a solution in works for this, but it would probably wait for 4.1. But the basic idea is that instead of always falling back onto the "default + project" theme combo (even in the editor) we'd be able to define parts of the scene tree that fall back on an arbitrary combination of global themes. So in the editor all controls would fall back onto the "default + editor" theme combo.

This would help for the cases you describe, as well as all the times the project theme "leaked" into the editor, and will be helpful for some other UX stuff too.

Reply:
> Previously tooltip popup was added as a child of `tooltip_serve_target` (previously `tooltip_owner`). It is now added as a child of the viewport for reuse.

And what prevents us from re-parenting it to the current tooltip owner instead?

Reply:
> And what prevents us from re-parenting it to the current tooltip owner instead?

Please see https://github.com/godotengine/godot/issues/59798#issuecomment-1246716710.

> In the user project you can the project theme for the tooltip styling, and this won't happen. The problem with the editor is quite specific to the way we have to configure its theme. But as mentioned above, I have a solid idea on how to resolve that for the editor.

We allow resetting the theme property for any control in the node path, so there is no guarantee that the user will not re-theme some part of the subtree (eg some desktop application development?).

Reply:
Due to delayed display and unstable property trees, this PR does not guarantee that the program will not crash (at least not increase the risk) when `Viewport::_gui_show_tooltip` is called. But it can prevent the program from crashing when `Viewport::_gui_cancel_tooltip` is called.

Reply:
> > And what prevents us from re-parenting it to the current tooltip owner instead?
> 
> Please see [#59798 (comment)](https://github.com/godotengine/godot/issues/59798#issuecomment-1246716710).

So you're saying that the problem is that it gets deleted when its parent is freed if we do that?

> We allow resetting the theme property for any control in the node path, so there is no guarantee that the user will not re-theme some part of the subtree (eg some desktop application development?).

Indeed, but what I mean is that it can be configured in projects to some degree without hacks, whereas in the editor we don't have this opportunity (yet).

Reply:
> > > And what prevents us from re-parenting it to the current tooltip owner instead?
> > 
> > 
> > Please see [#59798 (comment)](https://github.com/godotengine/godot/issues/59798#issuecomment-1246716710).
> 
> So you're saying that the problem is that it gets deleted when its parent is freed if we do that?

When the parent is freed, `Viewport::_gui_cancel_tooltip` may also be called and it gets deleted again.